src/inspector_agent.cc

```diff
@@ -520,49 +506,70 @@ void AgentImpl::WorkerRunIO() {
   uv_run(&child_loop_, UV_RUN_DEFAULT);
 }
 
+void AgentImpl::AppendMessage(MessageQueue* queue, int session_id,
+                              const String16& message) {
+  Mutex::ScopedLock scoped_lock(queue_lock_);
+  queue->push_back(std::make_pair(session_id, message));
+}
+
+void AgentImpl::SwapBehindLock(MessageQueue* vector1, MessageQueue* vector2) {
+  Mutex::ScopedLock scoped_lock(queue_lock_);
+  vector1->swap(*vector2);
+}
+
+void AgentImpl::PostIncomingMessage(const String16& message) {
+  AppendMessage(&incoming_message_queue_, frontend_session_id_, message);
+  v8::Isolate* isolate = parent_env_->isolate();
+  platform_->CallOnForegroundThread(isolate,
+    new DispatchOnInspectorBackendTask(this));
+  isolate ->RequestInterrupt(InterruptCallback, this);
+  uv_async_send(&data_written_);
+}
+
 void AgentImpl::OnInspectorConnectionIO(inspector_socket_t* socket) {
   if (client_socket_) {
     DisconnectAndDisposeIO(socket);
     return;
   }
   client_socket_ = socket;
-  inspector_read_start(socket, OnBufferAlloc, AgentImpl::OnRemoteDataIO);
-  platform_->CallOnForegroundThread(parent_env_->isolate(),
-      new SetConnectedTask(this, true));
+  inspector_read_start(socket, OnBufferAlloc, DataCallback);
+  frontend_session_id_++;
+  PostIncomingMessage(String16(TAG_CONNECT, sizeof(TAG_CONNECT) - 1));
 }
 
-void AgentImpl::PostMessages() {
+void AgentImpl::DispatchMessages() {
   if (dispatching_messages_)
     return;
   dispatching_messages_ = true;
-  std::vector<std::string> messages;
-  SwapBehindLock(&AgentImpl::message_queue_, &messages);
-  for (auto const& message : messages)
-    inspector_->dispatchMessageFromFrontend(
-        String16::fromUTF8(message.c_str(), message.length()));
+  MessageQueue tasks;
+  SwapBehindLock(&incoming_message_queue_, &tasks);
+  for (const std::pair<int, const String16>& pair : tasks) {
+    const String16& message = pair.second;
+    if (message == TAG_CONNECT) {
+      CHECK(!connected_);
```

Comment:
I expect that the linter will complain that you should use `CHECK_EQ(connected, false)` here.